go/pvlsource/pvlsource.go

```diff
@@ -0,0 +1,326 @@
+// Copyright 2017 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package pvlsource
+
+import (
+	"crypto/sha512"
+	"encoding/hex"
+	"encoding/json"
+	"io/ioutil"
+	"sync"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/keybase1"
+)
+
+// Older than this will try to refresh merkle root.
+// Measures time since merkle root fetched, not time since published.
+const tShouldRefresh time.Duration = 1 * time.Hour
+
+// Older than this is too old to use. All identifies will fail.
+// Measures time since merkle root fetched, not time since published.
+const tRequireRefresh time.Duration = 24 * time.Hour
+
+var dbKey = libkb.DbKey{
+	Typ: libkb.DBPvl,
+	Key: "active",
+}
+
+// Bump this to ignore existing cache entries.
+const dbVersion = 1
+
+type entry struct {
+	DBVersion int
+	Hash      string
+	PvlKit    string
+}
+
+// PvlSource is the way to get active pvl.
+// Talks to MerkleClient
+// Has an in-memory and LocalDB cache.
+type PvlSourceImpl struct {
+	libkb.Contextified
+	sync.Mutex
+
+	mem *entry
+}
+
+var _ libkb.PvlSource = (*PvlSourceImpl)(nil)
+
+// NewPvlSource creates a new source and installs it into G.
+func NewPvlSourceAndInstall(g *libkb.GlobalContext) libkb.PvlSource {
+	s := &PvlSourceImpl{
+		Contextified: libkb.NewContextified(g),
+	}
+	g.SetPvlSource(s)
+	return s
+}
+
+type pvlKitT struct {
+	KitVersion int                     `json:"kit_version"`
+	Ctime      int                     `json:"ctime"`
+	Tab        map[int]json.RawMessage `json:"tab"`
+}
+
+// Get PVL to use.
+func (s *PvlSourceImpl) GetPVL(ctx context.Context, pvlVersion int) (string, error) {
+
```

Comment:
Sure, now warns on errors here. Don't want to dump the whole json because it's big, but the hash is logged and that's enough for us to find the json.